tests/firedrake/regression/test_interpolation_operators.py

```diff
@@ -17,58 +18,88 @@ def params():
 
 @pytest.mark.parallel([1, 2])
 @pytest.mark.parametrize("value_shape", ["scalar", "vector"], ids=lambda v: f"fs_type={v}")
-@pytest.mark.parametrize("mat_type", [None, "aij", "baij"], ids=lambda v: f"mat_type={v}")
-def test_same_mesh_mattype(value_shape, mat_type):
+@pytest.mark.parametrize("mat_type", [None, "aij", "baij", "matfree"], ids=lambda v: f"mat_type={v}")
+@pytest.mark.parametrize("mode", ["forward", "adjoint"], ids=lambda v: f"mode={v}")
+def test_same_mesh_mattype(value_shape, mat_type, mode):
     if COMM_WORLD.size > 1:
         prefix = "mpi"
     else:
         prefix = "seq"
-    mesh = UnitSquareMesh(4, 4)
+
+    mesh = UnitSquareMesh(10, 10)
+    x, y = SpatialCoordinate(mesh)
     if value_shape == "scalar":
         fs_type = FunctionSpace
+        expr = x**2 + y**2
     else:
         fs_type = VectorFunctionSpace
+        expr = as_vector([x**2, y**2])
+
     V1 = fs_type(mesh, "CG", 1)
     V2 = fs_type(mesh, "CG", 2)
 
-    u = TrialFunction(V1)
+    if mode == "forward":
+        exact = Function(V1).interpolate(expr)
+        interp = interpolate(TrialFunction(V2), V1)  # V1 x V2^* -> R
+        f = Function(V2).interpolate(expr)
+    elif mode == "adjoint":
+        v1 = TestFunction(V1)
+        exact = assemble(inner(1, sum(v1) if value_shape == "vector" else v1) * dx)
+        interp = interpolate(TestFunction(V1), TrialFunction(V2.dual()))  # V2^* x V1 -> R
+        v2 = TestFunction(V2)
+        f = inner(1, sum(v2) if value_shape == "vector" else v2) * dx
 
-    interp = interpolate(u, V2)
     assert isinstance(get_interpolator(interp), SameMeshInterpolator)
-    res = assemble(interp, mat_type=mat_type)
 
-    if value_shape == "scalar":
+    I_mat = assemble(interp, mat_type=mat_type)
+    assert isinstance(I_mat, ImplicitMatrix if mat_type == "matfree" else Matrix)
+
+    if mat_type == "matfree":
+        assert I_mat.petscmat.type == "python"
+    elif value_shape == "scalar":
         # Always seqaij for scalar
-        assert res.petscmat.type == prefix + "aij"
+        assert I_mat.petscmat.type == prefix + "aij"
     else:
         # Defaults to seqaij
-        assert res.petscmat.type == prefix + (mat_type if mat_type else "aij")
+        assert I_mat.petscmat.type == prefix + (mat_type if mat_type else "aij")
+
+    res = assemble(action(I_mat, f))
+    assert np.allclose(res.dat.data, exact.dat.data)
```

Comment:
Test `assemble(adjoint(I_mat), fstar)`

Reply:
But python matrices support `hermitianTranspose` via tha MatType Transpose (which swaps the roles of mult and multTranspose)